poly/src/polynomial/mod.rs

```diff
@@ -1,164 +1,84 @@
-        } else if divisor.is_zero() {
-            panic!("Dividing by zero polynomial")
-        } else if self.degree() < divisor.degree() {
-            Some((DensePolynomial::zero(), self.clone().into()))
-        } else {
-            // Now we know that self.degree() >= divisor.degree();
-            let mut quotient = vec![F::zero(); self.degree() - divisor.degree() + 1];
-            let mut remainder: DensePolynomial<F> = self.clone().into();
-            // Can unwrap here because we know self is not zero.
-            let divisor_leading_inv = divisor.leading_coefficient().unwrap().inverse().unwrap();
-            while !remainder.is_zero() && remainder.degree() >= divisor.degree() {
-                let cur_q_coeff = *remainder.coeffs.last().unwrap() * divisor_leading_inv;
-                let cur_q_degree = remainder.degree() - divisor.degree();
-                quotient[cur_q_degree] = cur_q_coeff;
-
-                for (i, div_coeff) in divisor.iter_with_index() {
-                    remainder[cur_q_degree + i] -= &(cur_q_coeff * div_coeff);
-                }
-                while let Some(true) = remainder.coeffs.last().map(|c| c.is_zero()) {
-                    remainder.coeffs.pop();
-                }
-            }
-            Some((DensePolynomial::from_coefficients_vec(quotient), remainder))
-        }
-    }
-}
-impl<'a, F: 'a + FftField> DenseOrSparsePolynomial<'a, F> {
-    /// Construct `Evaluations` by evaluating a polynomial over the domain
-    /// `domain`.
-    pub fn evaluate_over_domain<D: EvaluationDomain<F>>(
-        poly: impl Into<Self>,
-        domain: D,
-    ) -> Evaluations<F, D> {
-        let poly = poly.into();
-        poly.eval_over_domain_helper(domain)
-    }
+    /// Returns the number of variables in `self`
+    fn num_vars(&self) -> usize;
 
-    fn eval_over_domain_helper<D: EvaluationDomain<F>>(self, domain: D) -> Evaluations<F, D> {
-        match self {
-            SPolynomial(Cow::Borrowed(s)) => {
-                let evals = domain.elements().map(|elem| s.evaluate(elem)).collect();
-                Evaluations::from_vec_and_domain(evals, domain)
-            }
-            SPolynomial(Cow::Owned(s)) => {
-                let evals = domain.elements().map(|elem| s.evaluate(elem)).collect();
-                Evaluations::from_vec_and_domain(evals, domain)
-            }
-            DPolynomial(Cow::Borrowed(d)) => {
-                Evaluations::from_vec_and_domain(domain.fft(&d.coeffs), domain)
-            }
-            DPolynomial(Cow::Owned(mut d)) => {
-                domain.fft_in_place(&mut d.coeffs);
-                Evaluations::from_vec_and_domain(d.coeffs, domain)
-            }
-        }
-    }
+    /// Outputs an `l`-variate polynomial which is the sum of `l` `d`-degree univariate
+    /// polynomials where each coefficient is sampled uniformly at random.
+    fn rand<R: Rng>(d: usize, num_vars: usize, rng: &mut R) -> Self;
```

Comment:
Maybe we should generalize this to take in the monomial structure of the polynomial? I.e. a list of terms?

Reply:
Hm I'm not quite sure I understand. What would be a use case where that might be useful?

Reply:
Added an issue for future resolution